### PR TITLE
[Draft] Add cupy support for sparse matrix cg sampling

### PR DESCRIPTION
**Pull request by chinandrew**
Add cupy support for the gibbs() function for sparse matrices using the cg sampling method. 

Currently contains the cupy operations to the reg_coef_sampler and cg_sampler modules since those are where the heavy linalg occurs. Theoretically could make it a bit more widespread but I think that'd add more complexity with marginal performance gains. 

Probably needs couple tests to verify that errors are properly handled and messaged for failed imports or not implemented sections, and then will remove the draft status.

### CODE REVIEW

File: bayesbridge/design_matrix/sparse_matrix.py [hunk 1]
```diff
@@ -69,35 +74,47 @@ def dot(self, v):
...
         return result
 
-    def main_Tdot(self, v):
-        X = self.X_main
-        result = mkl_csr_matvec(X, v, transpose=True) \
-            if self.use_mkl else X.T.dot(v)
-        result -= np.sum(v) * self.column_offset
+    def main_Tdot(self, v, use_cupy):
+        if use_cupy:
+            X = self.X_main_cp
```
Comment (aki-nishimura) [inline bayesbridge/design_matrix/sparse_matrix.py:110]: Not necessary, it seems.

File: bayesbridge/design_matrix/sparse_matrix.py [hunk 2]
```diff
@@ -159,3 +176,9 @@ def toarray(self):
 
     def extract_matrix(self, order=None):
         pass
+
+    def _allocate_cupy_matrix(self):
+        if not cupy:
```
Comment (aki-nishimura) [inline bayesbridge/design_matrix/sparse_matrix.py:181]: I prefer explicitly checking for `None`; using the "faultiness" of `None` can obscure the intention and be confusing to unexperienced Python programmers.

Comment (chinandrew) [reply bayesbridge/design_matrix/sparse_matrix.py]: :+1: updated

File: bayesbridge/design_matrix/sparse_matrix.py (hunk 1 above)
Comment (chinandrew) [reply bayesbridge/design_matrix/sparse_matrix.py]: :+1: fixed

**Comment by chinandrew** [comment]
> Did you forget `()`? Also, how about we make `self.model.design._allocate_cupy_matrix()` throw a `NotImplementedError` when the design matrix is dense? It is more explicit. That also avoids having to import `SparseDesignMatrix` just to type check, thereby breaking the modular design. (Though I probably break that elsewhere.)

Yup I forgot it :facepalm: . I like your suggestion too, I've updated the code now. Also the cupy option should be working with this latest commit. Would be nice to clean up some of the numerous `if self.use_cupy`, but it should be fully functional as is with a simple gibbs(..., use_cupy=True). The chain initialization is all done on CPU, and GPU is used for the cg sampling. 

On gi_bleed, loading the data takes 22 seconds, the chain initialization takes 102 seconds, and the gibbs iterations takes 32 seconds. There's a handful of odds and operations in there that take up <10 seconds, for a total run time of 2 min 44s.

**Comment by chinandrew** [comment]
Closing in favor of #16

**Pull request closed without merging**